botocore/configprovider.py

```diff
@@ -0,0 +1,419 @@
+        :param methods: A tuple of methods to allow in the chain. This is for
+            backwards compatibility, the tuple is of strs with the values
+            instance, env, config. Omitting any of those values will skip
+            adding the corresponding config loading method to the chain.
+        """
+        self._session = session
+        if environ is None:
+            environ = os.environ
+        self._environ = environ
+        self._methods = methods
+
+    def build_config_chain(self, instance=True, env_vars=None,
+                           config_property=None, default=None, cast=None):
+        """Build a config chain following the standard botocore pattern.
+
+        In botocore most of our config chains follow the the precendence:
+        session_instance_variables, environment, config_file, default_value.
+
+        This is a convenience function for creating a chain that follow
+        that precendence.
+
+        :type instance: bool
+        :param instance: This indicates whether or not session instance
+            variable lookup should be included in the config chain. By default
+            it is True.
+
+        :type env_vars: str or list of str or None
+        :param env_vars: One or more environment variable names to search for
+            this value. They are searched in order. If it is None it will
+            not be added to the chain.
+
+        :type config_property: str or None
+        :param config_property: The string name of the key in the config file
+            for this config option. If it is None it will not be added to the
+            chain.
+
+        :type default: Any
+        :param default: Any constant value to be returned. If this is a
+            callable it will be treated as a lazy value, and the callable will
+            be called when the value is needed.
+
+        :type cast: None or callable
+        :param cast: If this value is None then it has no affect on the return
+            type. Otherwise, it is treated as a function that will cast our
+            provided type.
+
+        :rvalue: ConfigChain
+        :returns: A ConfigChain that resolves in the order env_vars ->
+            config_property -> default. Any values that were none are
+            omitted form the chain.
+        """
+        providers = []
+        if instance:
+            providers.append(
+                LazyProvider(
+                    provider_function=lambda: OpenDictProvider(
+                        source=self._session.instance_variables(),
+                    ),
+                )
+            )
+        if env_vars is not None:
+            providers.append(
+                ClosedDictProvider(
+                    keys=env_vars,
+                    source=self._environ,
+                )
+            )
+        if config_property is not None:
+            providers.append(
+                LazyProvider(
+                    provider_function=lambda: ClosedDictProvider(
```

Comment:
It might make sense to rename `provider_function` to something like `provide_factory`. First reading this, I would have assumed we would have been providing a lazily called function (similar to `provide()`) that provides the value.